src/Interpreters/Cache/SLRUFileCachePriority.cpp

```diff
@@ -83,92 +97,187 @@ IFileCachePriority::IteratorPtr SLRUFileCachePriority::add( /// NOLINT
     }
+
+    return iterator;
 }
 
 bool SLRUFileCachePriority::collectCandidatesForEviction(
     size_t size,
+    size_t elements,
     FileCacheReserveStat & stat,
     EvictionCandidates & res,
     IFileCachePriority::IteratorPtr reservee,
-    FinalizeEvictionFunc & finalize_eviction_func,
     const UserID & user_id,
     const CachePriorityGuard::Lock & lock)
 {
-    /// If `it` is nullptr, then it is the first space reservation attempt
+    /// If `reservee` is nullptr, then it is the first space reservation attempt
     /// for a corresponding file segment, so it will be directly put into probationary queue.
     if (!reservee)
     {
-        return probationary_queue.collectCandidatesForEviction(size, stat, res, reservee, finalize_eviction_func, user_id, lock);
+        return probationary_queue.collectCandidatesForEviction(size, elements, stat, res, reservee, user_id, lock);
     }
 
-    /// If `it` not nullptr (e.g. is already in some queue),
+    auto * slru_iterator = assert_cast<SLRUIterator *>(reservee.get());
+    bool success = false;
+
+    /// If `reservee` is not nullptr (e.g. is already in some queue),
     /// we need to check in which queue (protected/probationary) it currently is
     /// (in order to know where we need to free space).
-    if (!assert_cast<SLRUIterator *>(reservee.get())->is_protected)
+    if (!slru_iterator->is_protected)
+    {
+        chassert(slru_iterator->lru_iterator.cache_priority == &probationary_queue);
+        success = probationary_queue.collectCandidatesForEviction(size, elements, stat, res, reservee, user_id, lock);
+    }
+    else
+    {
+        chassert(slru_iterator->lru_iterator.cache_priority == &protected_queue);
+        /// Entry is in protected queue.
+        /// Check if we have enough space in protected queue to fit a new size of entry.
+        /// `size` is the increment to the current entry.size we want to increase.
+        success = collectCandidatesForEvictionInProtected(size, elements, stat, res, reservee, user_id, lock);
+    }
+
+    /// We eviction_candidates (res) set is non-empty and
+    /// space reservation was successful, we will do eviction from filesystem
+    /// which is executed without a cache priority lock.
+    /// So we made space reservation from a certain queue (protected or probationary),
+    /// but we should remember that there is an increasePriority operation
+    /// which can be called concurrently to space reservation.
+    /// This operation can move elements from one queue to another.
+    /// We need to make sure that this does not happen for the elements
+    /// which are in the process of unfinished space reservation.
+    if (success && res.size() > 0)
     {
-        return probationary_queue.collectCandidatesForEviction(size, stat, res, reservee, finalize_eviction_func, user_id, lock);
+        slru_iterator->movable = false;
+        res.onFinalize([=](const CachePriorityGuard::Lock &){ slru_iterator->movable = true; });
```

Comment:
Is the order of `onFinalize` callbacks defined (it's not clear from the code)?
E.g., if `slru_iterator->movable = true;` is executed before the `downgrade` callback, the candidates will be placed into probationary queue.
But if its the other way around,`downgrade` will throw.

Reply:
> Is the order of onFinalize callbacks defined (it's not clear from the code)?

No order, as now it is not important.

> E.g., if slru_iterator->movable = true; is executed before the downgrade callback, the candidates will be placed into probationary queue.

Here the order should not affect anything for downgrade, this is explained in this comment https://github.com/ClickHouse/ClickHouse/blob/5100b0d52eaa61b458660f43fc8badcc31835d09/src/Interpreters/Cache/SLRUFileCachePriority.h#L130-L132 e.g. `movable` flag is needed only for `increasePriority`, but for `downgrade` it does not make any difference, because if we are currently reserving space for a file segment (for which we have marked `movable=false` before and set to `movable=true` in this callback) then this file segment is non-releasable (its ref count is > 1, because it is currently being used in some query), so we will not choose it as downgrade candidate. But I agree that it looks confusing...

Reply:
> because if we are currently reserving space for a file segment (for which we have marked movable=false before and set to movable=true in this callback) then this file segment is non-releasable (its ref count is > 1, because it is currently being used in some query), so we will not choose it as downgrade candidate.

It just occurred to me that there is not reason why we should not downgrade an element if it is non-releasable. But it is just how code of SLRU works right now because it uses `LRUFileCachePriority::collectCandidatesForEviction` underneath, which collects elements only if they are releasable. But we could just add a flag to this method to allow collecting releasable elements as well. So great you made this comment which set me thinking about this. Though I'd better make this change in a separate PR, then the order of finalize func's will be important indeed.